### PR TITLE
Simplify '.content' by removing CONTENT_MISSING

### LINKED ISSUE #963: Simplify `.content`
Our HTTPRequest/HTTPResponse's `.content` attribute can currently be either

- `None`, if we have not read the body yet (e.g. in the `reponseheaders` hook)
- `CONTENT_MISSING=0`, if we deliberately deleted the body
- `b""`, if we read an (empty) body.
- `b"some-content"`, if we actually read some content.

While the first three values intuitively evaluate to `False`, I feel the bytes/int/None heterogenity a bit confusing. I'd like to combine the first two cases, so that the type of this attribute can accurately be described as `Optional[bytes]`. This would fix #692 and make the code IMO easier to understand. We could also first deprecate `CONTENT_MISSING` gradually by setting it to `None`, and then remove it later on. @cortesi, would you be okay with this approach?

### PR DESCRIPTION
**Pull request by MatthewShao**
fixes #963
As @mhils said, I did this in two step:

- [x] setting ```None``` to ```CONTENT_MISSING``` (to see if anything go wrong)
- [x] replacing ```CONTENT_MISSING``` with ```None```

The relative comments and document have been updated too, hope I didn't miss anything.

### CODE REVIEW

File: mitmproxy/models/http.py [hunk 1]
```diff
@@ -226,9 +225,8 @@ class HTTPResponse(MessageMixin, Response):
 
         headers: Headers object
 
-        content: Content of the request, None, or CONTENT_MISSING if there
-        is content associated, but not present. CONTENT_MISSING evaluates
-        to False to make checking for the presence of content natural.
+        content: Content of the request, the value is None if there is content
```
Comment (Kriechi) [inline mitmproxy/models/http.py]: ...Content of the **response**...

File: mitmproxy/protocol/http.py [hunk 2]
```diff
@@ -50,8 +50,8 @@ def read_response_body(self, request, response):
         yield "this is a generator"  # pragma: no cover
 
     def send_response(self, response):
-        if response.content == CONTENT_MISSING:
-            raise HttpException("Cannot assemble flow with CONTENT_MISSING")
+        if response.content == None:
+            raise HttpException("Cannot assemble flow with None content")
```
Comment (Kriechi) [inline mitmproxy/protocol/http.py]: ...Cannot assemble flow with **missing content**

File: netlib/http/http1/assemble.py [hunk 3]
```diff
@@ -3,12 +3,10 @@
...
 import itertools
 from ...exceptions import HttpException
-from .. import CONTENT_MISSING
-
 
 def assemble_request(request):
-    if request.content == CONTENT_MISSING:
-        raise HttpException("Cannot assemble flow with CONTENT_MISSING")
+    if request.content == None:
+        raise HttpException("Cannot assemble flow with None content")
```
Comment (Kriechi) [inline netlib/http/http1/assemble.py]: ...**missing content**

File: netlib/http/http1/assemble.py [hunk 4]
```diff
@@ -21,8 +19,8 @@ def assemble_request_head(request):
 
 
 def assemble_response(response):
-    if response.content == CONTENT_MISSING:
-        raise HttpException("Cannot assemble flow with CONTENT_MISSING")
+    if response.content == None:
+        raise HttpException("Cannot assemble flow with None content")
```
Comment (Kriechi) [inline netlib/http/http1/assemble.py]: ...**missing content**

File: mitmproxy/console/common.py [hunk 5]
```diff
@@ -418,7 +417,7 @@ def format_flow(f, focus, extended=False, hostheader=False, marked=False):
     if f.response:
         if f.response.content:
             contentdesc = netlib.utils.pretty_size(len(f.response.content))
-        elif f.response.content == CONTENT_MISSING:
+        elif f.response.content is None:
             contentdesc = "[content missing]"
         else:
             contentdesc = "[no content]"
```
Comment (Kriechi) [inline mitmproxy/console/common.py]: Just for clarification:
The else-block is only reached if the content is basically an empty string?
Or is this code non-reachable now?

Comment (MatthewShao) [inline mitmproxy/console/common.py]: I think it is only reached if the content is ```""``` (empty string) or it is ```False```.

Comment (mhils) [inline mitmproxy/console/common.py]: This is reached if the content is `""`. The confusion is justified though - can we please do an explicit `content == ""` check?

File: mitmproxy/models/http.py (hunk 1 above)
Comment (mhils) [inline mitmproxy/models/http.py]: @Kriechi: ?

File: mitmproxy/web/app.py [hunk 6]
```diff
@@ -26,7 +25,7 @@ def _strip_content(flow_state):
                 continue
             if message["content"]:
                 message["contentLength"] = len(message["content"])
-            elif message["content"] == CONTENT_MISSING:
+            elif message["content"] is None:
```
Comment (mhils) [inline mitmproxy/web/app.py]: This `if elif else` can be simplified to `if else` by combining the cases where content is not None.

**Comment by mhils** [comment]
Other than my minor comments above, LGTM! :smiley:

File: mitmproxy/models/http.py (hunk 1 above)
Comment (Kriechi) [inline mitmproxy/models/http.py]: Ah yes - it's ok.
For some reason the Github diff view indicated a different class...
screen shot 2016-03-26 at 18 07 15

**Comment by Kriechi** [comment]
Thanks! :tada:

**Pull request merged**